planner/core/memtable_predicate_extractor.go

```diff
@@ -0,0 +1,200 @@
+// We must request all components in the cluster via HTTP API for retrieving
+// configurations and filter them by `type/address` columns.
+//
+// The purpose of defining a `MemTablePredicateExtractor` is to optimize this
+// 1. Define a `ClusterConfigTablePredicateExtractor`
+// 2. Extract the `type/address` columns on the logic optimizing stage and save them via fields.
+// 3. Passing the extractor to the `ClusterConfigReaderExec` executor
+// 4. Executor sends requests to the target components instead of all of the components
+type MemTablePredicateExtractor interface {
+	// extractCols from predicats and returns the remained predicates
+	Extract(*expression.Schema, []*types.FieldName, []expression.Expression) (remained []expression.Expression)
+}
+
+// extractHelper contains some common utililty functions for all extractor.
+// define a individual struct instead of a bunch of unexported functions
+// to avoid polluting the global scope of current package.
+type extractHelper struct{}
+
+func (helper extractHelper) extractColInConsExpr(extractCols map[int64]*types.FieldName, expr *expression.ScalarFunction) (string, []types.Datum) {
+	args := expr.GetArgs()
+	col, isCol := args[0].(*expression.Column)
+	if !isCol {
+		return "", nil
+	}
+	name, found := extractCols[col.UniqueID]
+	if !found {
+		return "", nil
+	}
+	// All expressions in IN must be a constant
+	// SELECT * FROM t1 WHERE c IN ('1', '2')
+	var results []types.Datum
+	for _, arg := range args[1:] {
+		constant, ok := arg.(*expression.Constant)
+		if !ok || constant.DeferredExpr != nil || constant.ParamMarker != nil {
+			return "", nil
+		}
+		results = append(results, constant.Value)
+	}
+	return name.ColName.L, results
+}
+
+func (helper extractHelper) extractColEqConsExpr(extractCols map[int64]*types.FieldName, expr *expression.ScalarFunction) (string, []types.Datum) {
+	args := expr.GetArgs()
+	var col *expression.Column
+	var colIdx int
+	// c = 'rhs'
+	// 'lhs' = c
+	for i := 0; i < 2; i++ {
+		var isCol bool
+		col, isCol = args[i].(*expression.Column)
+		if isCol {
+			colIdx = i
+			break
+		}
+	}
+	if col == nil {
+		return "", nil
+	}
+
+	name, found := extractCols[col.UniqueID]
+	if !found {
+		return "", nil
+	}
+	// The `lhs/rhs` of EQ expression must be a constant
+	// SELECT * FROM t1 WHERE c='rhs'
+	// SELECT * FROM t1 WHERE 'lhs'=c
+	constant, ok := args[1-colIdx].(*expression.Constant)
+	if !ok || constant.DeferredExpr != nil || constant.ParamMarker != nil {
+		return "", nil
+	}
+	return name.ColName.L, []types.Datum{constant.Value}
+}
+
+func (helper extractHelper) intersection(lhs set.StringSet, datums []types.Datum, toLower bool) set.StringSet {
+	tmpNodeTypes := set.NewStringSet()
+	for _, datum := range datums {
+		var s string
+		if toLower {
+			s = strings.ToLower(datum.GetString())
+		} else {
+			s = datum.GetString()
+		}
+		tmpNodeTypes.Insert(s)
+	}
+	if len(lhs) > 0 {
+		return lhs.Intersection(tmpNodeTypes)
+	}
+	return tmpNodeTypes
```

Comment:
It is pretty strange that intersection of empty set and another set is not empty.